src/core/net/ip4_headers.hpp

```diff
@@ -0,0 +1,470 @@
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for IPv4 packet processing.
+ */
+
+#ifndef IP4_HEADERS_HPP_
+#define IP4_HEADERS_HPP_
+
+#include "openthread-core-config.h"
+
+#include <stddef.h>
+
+#include "common/clearable.hpp"
+#include "common/encoding.hpp"
+#include "common/message.hpp"
+#include "net/ip4_address.hpp"
+#include "net/ip6.hpp"
+#include "net/netif.hpp"
+#include "net/socket.hpp"
+
+namespace ot {
+
+/**
+ * @namespace ot::Ip4
+ *
+ * @brief
+ *   This namespace includes definitions for IPv4 networking.
+ *
+ */
+namespace Ip4 {
+
+using Encoding::BigEndian::HostSwap16;
+using Encoding::BigEndian::HostSwap32;
+
+using Ecn = Ip6::Ecn;
+
+/**
+ * @addtogroup core-ipv4
+ *
+ * @brief
+ *   This module includes definitions for the IPv4 network layer.
+ *
+ */
+
+/**
+ * @addtogroup core-ip4-ip4
+ *
+ * @brief
+ *   This module includes definitions for IPv4 networking used by NAT64.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This class implements IPv4 header generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class Header : public Clearable<Header>
+{
+public:
+    static constexpr uint8_t kVersionIHLOffset         = 0;
+    static constexpr uint8_t kTrafficClassOffset       = 1;
+    static constexpr uint8_t kTotalLengthOffset        = 2;
+    static constexpr uint8_t kIdenficationOffset       = 4;
+    static constexpr uint8_t kFlagsFragmentOffset      = 6;
+    static constexpr uint8_t kTTLOffset                = 8;
+    static constexpr uint8_t kProtocolOffset           = 9;
+    static constexpr uint8_t kHeaderChecksumOffset     = 10;
+    static constexpr uint8_t kSourceAddressOffset      = 12;
+    static constexpr uint8_t kDestinationAddressOffset = 16;
+
+    /**
+     * This method indicates whether or not the header appears to be well-formed.
+     *
+     * @retval TRUE    If the header appears to be well-formed.
+     * @retval FALSE   If the header does not appear to be well-formed.
+     *
+     */
+    bool IsValid(void) const;
+
+    /**
+     * This method initializes the Version to 6 and sets Traffic Class and Flow fields to zero.
+     *
+     * The other fields in the IPv4 header remain unchanged.
+     *
+     */
+    void InitVersionIHL(void) { SetVersionIHL(kVersIHLInit); }
```

Comment:
Use `Ihl` instead of IHL... (we treat abbreviation as noun and use `CamelCase`).
```suggestion
    void InitVersionIhl(void) { SetVersionIHl(kVersIhlInit); }
```